opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -1801,6 +1854,152 @@ namespace Opm {
+    data::WellBlockAveragePressures
+    BlackoilWellModel<TypeTag>::
+    computeWellBlockAveragePressures() const
+    {
+        auto wbpResult = data::WellBlockAveragePressures{};
+
+        using Calculated = PAvgCalculator::Result::WBPMode;
+        using Output = data::WellBlockAvgPress::Quantity;
+
+        this->wbpCalculationService_.collectDynamicValues();
+
+        const auto numWells = this->wells_ecl_.size();
+        for (auto wellID = 0*numWells; wellID < numWells; ++wellID) {
+            const auto calcIdx = this->wbpCalcMap_[wellID].wbpCalcIdx_;
+            const auto& well = this->wells_ecl_[wellID];
+
+            if (! well.hasRefDepth()) {
+                // Can't perform depth correction without at least a
+                // fall-back datum depth.
+                continue;
+            }
+
+            this->wbpCalculationService_
+                .inferBlockAveragePressures(calcIdx, well.pavg(),
+                                            this->gravity_,
+                                            well.getWPaveRefDepth());
+
+            const auto& result = this->wbpCalculationService_
+                .averagePressures(calcIdx);
+
+            auto& reported = wbpResult.values[well.name()];
+
+            reported[Output::WBP]  = result.value(Calculated::WBP);
+            reported[Output::WBP4] = result.value(Calculated::WBP4);
+            reported[Output::WBP5] = result.value(Calculated::WBP5);
+            reported[Output::WBP9] = result.value(Calculated::WBP9);
+        }
+
+        return wbpResult;
+    }
+
+
+
+
+
+    template <typename TypeTag>
+    ParallelWBPCalculation::EvaluatorFactory
+    BlackoilWellModel<TypeTag>::
+    makeWellSourceEvaluatorFactory(const std::vector<Well>::size_type wellIdx) const
+    {
+        using Span = PAvgDynamicSourceData::SourceDataSpan<double>;
+        using Item = typename Span::Item;
+
+        return [wellIdx, this]() -> ParallelWBPCalculation::Evaluator
+        {
+            if (! this->wbpCalcMap_[wellIdx].openWellIdx_.has_value()) {
+                // Well is stopped/shut.  Return evaluator for stopped wells.
+                return []([[maybe_unused]] const int connIdx, Span sourceTerm)
+                {
+                    // Well/connection is stopped/shut.  Set all items to
+                    // zero.
+
+                    sourceTerm
+                        .set(Item::Pressure      , 0.0)
+                        .set(Item::PoreVol       , 0.0)
+                        .set(Item::MixtureDensity, 0.0);
+                };
+            }
+
+            // Well is open.  Return an evaluator for open wells/open connections.
+            return [this, wellPtr = this->well_container_[*this->wbpCalcMap_[wellIdx].openWellIdx_].get()]
+                (const int connIdx, Span sourceTerm)
+            {
+                // Note: The only item which actually matters for the WBP
+                // calculatoion at the well reservoir connection level is
```

Comment:
calculation

Reply:
> calculation

:+1: 

I've pushed an update to fix that.